---
title: 'Explore 1000 genomics data using unsupervised learning methods'
output: 
  github_document
editor_options: 
  markdown: 
    wrap: sentence
knitr:
  opts_chunk:
    set:
      cache: TRUE       # turn on caching globally
      cache.path: cache/  # optional: where to store cache files
bibliography: references.bib
---

```{r include = FALSE }
knitr::opts_chunk$set(echo=TRUE, warning=TRUE, message=TRUE)
knitr::opts_chunk$set(fig.width=6, fig.height=4) 
```

# Introduction

The 1000 Genomes Project is a landmark international collaboration that produced the most comprehensive public catalogue of human genetic variation to date.
It contains the genome sequencing of 2,504 individuals from 26 diverse populations around the world.
All data are freely available on the official website, providing an essential baseline for studies of population genetics, disease‚Äêassociation mapping, and functional genomics.
In this project, we explore the genetic dataset using unsupervising learning methods.

## Load data and packages

### Install & load required libraries

The following code chunk will load required libraries.
If you don't already have these installed, you'll first need to install them (recommended way is to use `BiocManager::install("packageName")`).

```{r, eval = FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("gdsfmt")
BiocManager::install("SNPRelate")


```

```{r, warning=FALSE}
library(gdsfmt) # provides the genomic data structure (GDS) file format for array-oriented bioinformatic data, which is a container for storing annotation data and SNP genotypes
library(SNPRelate) #Read SNP data and perform PCA
library(ggplot2) 
library(dplyr)
library(kableExtra) 
library(Rtsne) #t-SNE
library(pheatmap)
library(cluster)
theme_set(theme_bw()) # prettier ggplot plots 
```

### Download and read the data into R

I obtained the genetic data from 1000 genomic, followed the tutorial posted here (<https://www.biostars.org/p/335605/>).
LD-pruning was performed using the PLINK software.
SNPs with MAF \< 0.1 were removed.
After filtering, the data contain 451,115 SNPs from 2,504 samples.

```{r, eval = FALSE}
# Convert the plink format to GDS format used for the package. 
snpgdsBED2GDS(bed.fn = "data/1000genomicMerge.bed",
              bim.fn = "data/1000genomicMerge.bim",
              fam.fn = "data/1000genomicMerge.fam",
              out.gdsfn = "data/1000genomic.gds")

```

```{r}
genofile <- snpgdsOpen("data/1000genomic.gds") #read the genotype file 
genofile
PED <- read.table('data/20130606_g1k.ped', header = TRUE, skip = 0, sep = '\t') #Read in the population information file 

geno <- snpgdsGetGeno(genofile, snpfirstdim = FALSE, with.id = TRUE)
```

```{r}
tab <- data.frame(sample.id = geno$sample.id)
tab <- left_join(tab, PED, by = c("sample.id" = "Individual.ID"))
# Assigning the superpopulation group by browsering information from: http://www.internationalgenome.org/category/population/
tab <- tab %>%
  mutate(superpop = case_when(
    Population %in% c("ESN", "MSL", "YRI", "LWK", "GWD") ~ "African",
    Population %in% c("ASW", "ACB", "MXL", "PUR", "CLM", "PEL") ~ "Latin American",
    Population %in% c("CEU", "TSI", "GBR", "FIN", "IBS") ~ "European",
    Population %in% c("CHB", "JPT", "CHS", "CDX", "KHV", "CHD") ~ "East Asian", 
    Population %in% c("GIH", "PJL", "BEB", "STU", "ITU") ~ "South Asian"))  %>%
  mutate(superpop = as.factor(superpop))
table(tab$superpop) 
geno$superpop <- tab$superpop
geno$population <- tab$Population
```

The table lists the number of each population, where the detailed description can be found on the [website]((http://www.internationalgenome.org/category/population/)).

# Dimension reduction

## PCA

The common way to explore high-dimensional data is by using principal component analysis(PCA).
PCA is a dimensional reduction technique which preserves the dimensions with the largest variance in the data.
It is a useful initial means of finding hidden structures in the data, and also to determine the important sources of variance.

```{r, cache = TRUE}
pca <- snpgdsPCA(genofile, num.thread=8) #Perform the principal component analysis
```

```{r}
plot(pca$varprop[1:10], xlab = "Principal componets", ylab = "percentage of variance explained")
```

The plot shows the percentage of variance explained by the top 10 principal components.
Here we see the first 4 components almost explain all the genetic variance.

```{r}
tab <- data.frame(sample.id = pca$sample.id,
    EV1 = pca$eigenvect[,1],    # the first eigenvector
    EV2 = pca$eigenvect[,2],    # the second eigenvector
    EV3 = pca$eigenvect[,3], 
    EV4 = pca$eigenvect[,4], 
    stringsAsFactors = FALSE)
tab <- left_join(tab,PED, by = c("sample.id" = "Individual.ID"))

# Assigning the superpopulation group by browsering information from: http://www.internationalgenome.org/category/population/
tab <- tab %>%
  mutate(superpop = case_when(
    Population %in% c("ESN", "MSL", "YRI", "LWK", "GWD") ~ "African",
    Population %in% c("ASW", "ACB", "MXL", "PUR", "CLM", "PEL") ~ "Latin American",
    Population %in% c("CEU", "TSI", "GBR", "FIN", "IBS") ~ "European",
    Population %in% c("CHB", "JPT", "CHS", "CDX", "KHV", "CHD") ~ "East Asian", 
    Population %in% c("GIH", "PJL", "BEB", "STU", "ITU") ~ "South Asian"))  %>%
  mutate(superpop = as.factor(superpop))
```

```{r}
# Draw
plot(tab$EV1, tab$EV2, col=as.integer(tab$superpop), xlab="First component", ylab="Second component" )
legend("topleft", legend=levels(tab$superpop), pch="o", col=1:nlevels(tab$superpop), cex = 0.7)

plot(tab$EV1, tab$EV3, col=as.integer(tab$superpop), xlab="First component", ylab="Third component" )
legend("topleft", legend=levels(tab$superpop), pch="o", col=1:nlevels(tab$superpop), cex = 0.7)

lbls <- paste("PC", 1:4, "\n", format(pca$varprop[1:4] * 100, digits=2), "%", sep="")
pairs(pca$eigenvect[,1:4], col=tab$superpop, labels=lbls)
```

We observe that the superpopulations are mostly clustered in groups when visualizing the first and the third principal components.
We also observe that there is a wider spread of the first component among Latin Americans.
Some are more similar to European, while some are closer to African.
This can be attributed to greater genetic diversity among Latin Americans due to historical migration and population mixing.

```{r}
tab_Latin <- subset(tab, superpop == "Latin American") %>%
  mutate(Population = as.factor(Population))

plot(tab_Latin$EV1, tab_Latin$EV2, col=as.integer(tab_Latin$Population), xlab="First component", ylab="Second component")
legend("bottomleft", legend=levels(tab_Latin$Population), pch="o", col=1:nlevels(tab_Latin$Population), cex = 0.7)
```

Given that ASW are of African Ancestry in the Southwest US and ACB are of African Caribbean in Barbados, it is unsurprising to see the overlaps with African superpopulation from the PCA plot.
We also found some overlaps between PUR(Puerto Rican), CLM(Colombina) and European superpopulation.

```{r}
tab_EUR <- subset(tab, superpop == "European") %>%
  mutate(Population = as.factor(Population))

plot(tab_EUR$EV1, tab_EUR$EV2, col=as.integer(tab_EUR$Population), xlab="First component", ylab="Second component")
legend("topleft", legend=levels(tab_EUR$Population), pch="o", col=1:nlevels(tab_EUR$Population), cex = 0.7)


```

In the European population, the first component reflects geography across Europe, arranged from south to north.
The Iberian populations in Spain (IBS) exhibit the smallest first component, followed by the Tuscans in Italy (TSI), the British in England and Scotland (GBR), the residents of Utah with Northern and Western European ancestry (CEU), and the Finnish people (FIN) in Finland.
This replicates earlier results reported by [@novembre2008], which displayed a similar pattern.

## t-SNE

Another method that is an alternative to PCA is t-SNE.
In contrast to PCA, it accounts for non-linear interactions between the features.

```{r, cache = TRUE}
tSNE <- Rtsne(geno$genotype,verbose = TRUE,partial_pca = TRUE) 
```

```{r}
as.data.frame(tSNE$Y) %>% 
  mutate(population = geno$population) %>% 
  mutate(superpop = geno$superpop) %>%
  ggplot(aes(x = V1, y = V2, colour = superpop)) + 
    geom_point() + 
    xlab("tsne 1") +
    ylab("tsne 2") +
    ggtitle("tSNE")
```

It does not really separate the superpopulation into clusters.
It is not actually perform better than PCA.

# Clustering

## Hierarchical Clustering

Perform clustering based on the top 10 principal components of genetic matrix.

```{r}
distance <- dist(pca$eigenvect[, 1:10], method = 'euclidean') #Compute the Euclidean distance between each samples 
str(distance)
```

```{r, include = FALSE}
index <- sample(1:nrow(geno$genotype), 100)
geno_subset <- geno$genotype[index, ]
distance <- dist(geno_subset, method = 'manhattan') #Compute the manhattan distance of highly variable SNPs between each samples
```

```{r, include = FALSE}
snp_vars <- apply(geno$genotype, 2, var)
quantile(snp_vars, probs = 0.99)
index <- which(snp_vars > quantile(snp_vars, probs = 0.999))
geno_subset <- geno$genotype[, index]
distance <- dist(geno_subset, method = 'manhattan') #Compute the manhattan distance of highly variable SNPs between each samples. 
```

then, compute hierarchical clustering using 4 different linkage types: complete, single, average and Wald, and plot them.

```{r}
pr.hc.s <- hclust(distance, method = 'single')
pr.hc.c <- hclust(distance, method = 'complete')
pr.hc.a <- hclust(distance, method = 'average')
pr.hc.w <- hclust(distance, method = 'ward.D')

# plot them
op <- par(mar = c(0,4,4,2), mfrow = c(2,2))

plot(pr.hc.s, labels = FALSE, main = "Single")
plot(pr.hc.c, labels = FALSE, main = "Complete")
plot(pr.hc.a, labels = FALSE, main = "Average")
plot(pr.hc.w, labels = FALSE, main = "Ward")

par(op)
```

Find the 5 clusters based on the Ward linkage method.

```{r}
cluster_samples = cutree(pr.hc.w, k = 5)
table(cluster_samples, geno$superpop)

```

Cluster 3 contains only the East Asian population, and cluster 5 contains only the South Asian population.
The other clusters contain the mixture.
Let us look at the five clusters defined by the single linkage.

```{r}
cluster_samples = cutree(pr.hc.s, k = 5)
table(cluster_samples, geno$superpop)
```

The clustering based on the single linkage put most samples to the first cluster.

```{r}
cluster_samples = cutree(pr.hc.a, k = 5) 
table(cluster_samples, geno$superpop)
```

So is the average method.
We can also visualize the genetic distribution with heatmap

```{r}
# set pheatmap clustering parameters
clust_dist_col = "euclidean" 
clust_method = "ward.D"
clust_scale = "none"

## the annotation option uses the covariate object (pData(geo_obj)). It must have the same rownames, as the colnames in our data object (expr_scaled).  

pheatmap(t(pca$eigenvect[,1:10]), 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         scale = clust_scale, 
         clustering_method = clust_method, 
         clustering_distance_cols = clust_dist_col, 
         show_colnames = FALSE, show_rownames = FALSE,  
         main = "Clustering heatmap for top 10 principal components", 
         annotation = t(geno$superpop))
```

## K-mean clustering

Keep in mind that k-means makes certain assumptions about the data that may not always hold:\
- Variance of distribution of each variable is spherical\
- All variables have the same variance\
- A prior probability that all k clusters have the same number of members

One can argue that for genetic data, the assumption that all SNPs have the same variance is the most easily violated.

```{r}
set.seed(17)
k <- 5
pr.km <- kmeans(pca$eigenvect[, 1:10], centers = k, nstart =  50)

#We can look at the size of each cluster
pr.km$size

table(pr.km$cluster, geno$superpop)
```

K-means clustering successfully identifies the clusters of East Asian, European and South Asian populations.

## PAM algorithm

In K-medoids clustering, K representative objects (medoids) are chosen as cluster centers and objects are assigned to the center (medoid = cluster) with which they have minimum dissimilarity (Kaufman and Rousseeuw, 1990).\
Nice features of partitioning around medoids (PAM) are: (a) it accepts a dissimilarity (distance) matrix (use `diss = TRUE`), (b) it is more robust to outliers as the centroids of the clusters are data objects, unlike k-means.

Here we run PAM with k = 5.

```{r}
pr.pam <- pam(distance, k = 5)
table(pr.pam$clustering, geno$superpop)
```

We will now determine the optimal number of clusters, by looking at the average silhouette value.
This is a statistic introduced in the PAM algorithm, which lets us identify a suitable k.

**The silhouette plot** The `cluster` package contains the function `silhouette()` that compares the minimum average dissimilarity (distance) of each object to other clusters **with** the average dissimilarity to objects in its own cluster.
The resulting measure is called the "width of each object's silhouette".
A value close to 1 indicates that the object is similar to objects in its cluster compared to those in other clusters.
Thus, the average of all objects silhouette widths gives an indication of how well the clusters are defined.

```{r}
op <- par(mar = c(5,1,4,4))
plot(pr.pam, main = "Silhouette Plot for 5 clusters")
par(op)
```

We see that cluster 5 is the best defined cluster, followed by cluster 4.
We now draw a plot showing number of clusters in the x-axis and average silhouette widths in the y-axis, and determine the optimal choice for the number of clusters.

```{r silplot, cache = TRUE}
j <- 2:30
avg_width <- numeric(length(j))  # pre-allocate numeric vector
distance <- dist(pca$eigenvect[, 1:10])

for (i in j) {
  pam_result <- pam(distance, k = i , diss = TRUE)  # ensure 'distance' is a dissimilarity matrix
  avg_width[i - 1] <- pam_result$silinfo$avg.width
}

plot(j, avg_width, type = "b", pch = 19, col = "blue",
     xlab = "Number of clusters (k)",
     ylab = "Average silhouette width",
     main = "Silhouette Width vs. Number of Clusters")
```

It seems like the optimal number is 10.

# Discussion

In this project, I utilized various dimensionality reduction methods to visualize genetic data.
Principal Component Analysis (PCA) allows for the separation of samples based on their continental ancestry groups.
I also reproduced the finding that the first principal component distinguishes between northern and southern geographical locations within the European population.
Additionally, I identified that Latin Americans exhibit significantly greater genetic variance compared to other populations.
Furthermore, I applied several clustering algorithms to compare their performance on the genetic data.

# References
